File: quickstart/commands/ping.py
"""qs ping — mTLS GET request to verify the server is up and running."""

from __future__ import annotations

import argparse
import ssl
import sys
import urllib.request
from pathlib import Path

from quickstart import runner
from quickstart.config import ProjectConfig


def run_ping(args: argparse.Namespace, config: ProjectConfig) -> int:
    cert_dir = Path(args.certs) if args.certs else config.repo_root / "certs" / "devices" / "test-device"
    ca = Path(args.ca) if args.ca else config.repo_root / "certs" / "ca" / "ca.pem"

    client_cert = cert_dir / "client.pem"
    client_key = cert_dir / "client.key"

    for path, label in [(client_cert, "client cert"), (client_key, "client key"), (ca, "CA cert")]:
        if not path.exists():
            runner.error(f"{label} not found at {path}")
            return 1

    ctx = ssl.create_default_context(cafile=str(ca))
    ctx.load_cert_chain(certfile=str(client_cert), keyfile=str(client_key))

    url = f"https://{args.host}:{args.port}/"
    runner.step(f"Pinging {url}")

    if args.dry_run:
        print(f"  [dry-run] GET {url}")
        return 0

    try:
        req = urllib.request.Request(url, method="GET")
        with urllib.request.urlopen(req, context=ctx, timeout=10) as resp:
            body = resp.read().decode()
            runner.success(f"Status: {resp.status}")
            print(body)
            return 0
    except urllib.error.HTTPError as e:
        # Server responded — it's up, even if the status isn't 2xx
        runner.success(f"Status: {e.code}")
        print(e.read().decode())
        return 0
    except Exception as e:
        runner.error(f"Error: {e}")
        return 1


def register_subparser(sub: argparse._SubParsersAction) -> None:
    p = sub.add_parser("ping", help="mTLS GET to verify the server is up")
    p.add_argument("--certs", default=None, help="Directory with client.pem and client.key (default: certs/devices/test-device)")
    p.add_argument("--ca", default=None, help="CA certificate (default: certs/ca/ca.pem)")
    p.add_argument("--host", default="localhost", help="Server host (default: localhost)")
    p.add_argument("--port", type=int, default=443, help="Server port (default: 443)")

File: cli/src/commands/ping.rs
use std::path::{Path, PathBuf};

use ureq::tls::{Certificate, ClientCert, PrivateKey, RootCerts, TlsConfig};

use crate::config::ProjectConfig;
use crate::error::CliError;
use crate::runner;

/// Arguments for the `qs ping` command.
pub struct PingArgs {
    /// Optional path to the device certs directory (default: `certs/devices/test-device`).
    pub certs: Option<PathBuf>,
    /// Optional CA cert path (default: WebPki roots).
    pub ca: Option<PathBuf>,
    /// Target hostname.
    pub host: String,
    /// Target HTTPS port.
    pub port: u16,
    /// Print commands without executing.
    pub dry_run: bool,
}

/// Send an mTLS GET request to the target endpoint and print the response.
pub fn run_ping(args: &PingArgs, config: &ProjectConfig) -> Result<i32, CliError> {
    let cert_dir = args
        .certs
        .clone()
        .unwrap_or_else(|| config.repo_root.join("certs/devices/test-device"));
    let ca = args.ca.clone();

    let client_cert = cert_dir.join("client.pem");
    let client_key = cert_dir.join("client.key");

    for (path, label) in [(&client_cert, "client cert"), (&client_key, "client key")] {
        if !path.exists() {
            runner::error(&format!("{} not found at {}", label, path.display()));
            return Ok(1);
        }
    }

    if let Some(ref ca_path) = ca {
        if !ca_path.exists() {
            runner::error(&format!("CA cert not found at {}", ca_path.display()));
            return Ok(1);
        }
    }

    let url = format!("https://{}:{}/", args.host, args.port);
    runner::step(&format!("Pinging {}", url));

    if args.dry_run {
        println!("  [dry-run] GET {}", url);
        return Ok(0);
    }

    let agent = build_mtls_agent(ca.as_deref(), &client_cert, &client_key)?;

    match agent.get(&url).call() {
        Ok(response) => {
            let status = response.status();
            let body = response.into_body().read_to_string().unwrap_or_default();
            runner::success(&format!("Status: {}", status));
            println!("{}", body);
            Ok(0)
        }
        Err(ureq::Error::StatusCode(status)) => {
            runner::success(&format!("Status: {}", status));
            Ok(0)
        }
        Err(e) => {
            runner::error(&format!("Error: {}", e));
            Ok(1)
        }
    }
}

/// Build a ureq Agent configured for mTLS.
/// Uses WebPki roots by default; if `ca_path` is provided, uses that CA exclusively.
pub fn build_mtls_agent(
    ca_path: Option<&Path>,
    cert_path: &Path,
    key_path: &Path,
) -> Result<ureq::Agent, CliError> {
    let cert_pem = std::fs::read(cert_path)?;
    let client_cert = Certificate::from_pem(&cert_pem)
        .map_err(|e| CliError::Config(format!("failed to parse client cert: {}", e)))?;

    let key_pem = std::fs::read(key_path)?;
    let client_key = PrivateKey::from_pem(&key_pem)
        .map_err(|e| CliError::Config(format!("failed to parse client key: {}", e)))?;

    let root_certs = match ca_path {
        Some(path) => {
            let ca_pem = std::fs::read(path)?;
            let ca_cert = Certificate::from_pem(&ca_pem)
                .map_err(|e| CliError::Config(format!("failed to parse CA cert: {}", e)))?;
            RootCerts::Specific(std::sync::Arc::new(vec![ca_cert]))
        }
        None => RootCerts::WebPki,
    };

    let tls_config = TlsConfig::builder()
        .root_certs(root_certs)
        .client_cert(Some(ClientCert::new_with_certs(&[client_cert], client_key)))
        .build();

    let config = ureq::Agent::config_builder().tls_config(tls_config).build();

    Ok(ureq::Agent::new_with_config(config))
}

/// Build a plain ureq Agent with no mTLS.
pub fn build_plain_agent() -> ureq::Agent {
    ureq::Agent::new_with_defaults()
}
